### PR TITLE
include time series embedding from https://arxiv.org/abs/1901.10738

### PR DESCRIPTION
**Pull request by vafl**
*Description of changes:*

This includes a modified version of the code from https://arxiv.org/abs/1901.10738.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by kashif** [comment]
@vafl could you also try the barlow twins loss as well? https://arxiv.org/abs/2103.03230

**Comment by vafl** [comment]
> https://arxiv.org/abs/2103.03230

Cool, thanks for the suggestion @kashif ! I'll try this out.

**Comment by kashif** [comment]
@vafl awesome... I wanted to add some time series augmentation transforms, so hopefully, I can share something soon.

File: src/gluonts/nursery/ts_embeddings/causal_cnn.py [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+# SOFTWARE.
+
+
+import torch
+import torch.nn.functional as F
+
+
+class Chomp1d(torch.nn.Module):
+    """
+    Removes the last elements of a time series.
```
Comment (mbohlkeschneider) [inline src/gluonts/nursery/ts_embeddings/causal_cnn.py:48]: I think we converged on `Numpy` style docstrings. Maybe we want to make this consistent?

File: src/gluonts/nursery/ts_embeddings/causal_cnn.py [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+
+    Takes as input a three-dimensional tensor (`B`, `C`, `L`) where `B` is the
+    batch size, `C` is the number of input channels, and `L` is the length of
+    the input. Outputs a three-dimensional tensor (`B`, `C`, `L - s`) where `s`
+    is the number of elements to remove.
+
+    @param chomp_size Number of elements to remove.
+    """
+
+    def __init__(self, chomp_size):
```
Comment (mbohlkeschneider) [inline src/gluonts/nursery/ts_embeddings/causal_cnn.py:58]: Should we add type annotations?

**Review by mbohlkeschneider — CHANGES_REQUESTED** [review]
Maybe adding a smoke test for the model would be good?

File: src/gluonts/nursery/ts_embeddings/ShowEmbeddings.ipynb [hunk 3]
```diff
@@ -0,0 +1,158 @@
+{
+ "cells": [
+  {
+   "cell_type": "code",
+   "execution_count": null,
```
Comment (timoschowski) [inline src/gluonts/nursery/ts_embeddings/ShowEmbeddings.ipynb:5]: One of @lostella 's suggestions was to move the notebooks to the examples folder.

**Review by timoschowski — COMMENTED** [review]
Looks good to me, thanks @vafl . A minor comment from me on the location of the notebook (and then @mbohlkeschneider 's comments.

**Comment by kashif** [comment]
BTW here is barlow twins integrated with your code base if you haven't implemented it yet, else please ignore: https://github.com/vafl/gluon-ts/pull/1

File: src/gluonts/nursery/ts_embeddings/ShowEmbeddings.ipynb (hunk 3 above)
Comment (lostella) [reply src/gluonts/nursery/ts_embeddings/ShowEmbeddings.ipynb:5]: @jaheba pointed out that jupyter notebooks may not end up installed with the package, so maybe this is not really urgent

File: src/gluonts/nursery/ts_embeddings/embed_model.py [hunk 4]
```diff
@@ -240,7 +256,11 @@ def training_step(self, batch, batch_idx):
             prog_bar=False,
             logger=True,
         )
-        return loss
+        # the loss can become NaN if we pick windows that only contain NaN values. We'll ignore these.
+        if torch.any(loss.isnan()):
```
Comment (timoschowski) [inline src/gluonts/nursery/ts_embeddings/embed_model.py:260]: I'm not sure if this is a good idea. Could you take a look at this and advice @vafl ?

**Review by timoschowski — COMMENTED** [review]
This is a question for feedback that I have on my own code for @vafl

**Comment by kashif** [comment]
@timoschowski and @vafl In the mean time i have MoCo working as well if you would like to have that let me know... (I needed to refactor the code a bit for that since it needs two copies of the encoder network etc.)

**Comment by timoschowski** [comment]
awesome @kashif . I don't know what the best procedure is here, whether to do another PR or add your commits to this PR? I any case, I think we should merge this soon @vafl ?

**Comment by timoschowski** [comment]
> @timoschowski and @vafl In the mean time i have MoCo working as well if you would like to have that let me know... (I needed to refactor the code a bit for that since it needs two copies of the encoder network etc.)

Can you push this onto the PR @kashif  directly? I'm getting a bit lost with the PRs otherwise. And sorry that I pushed your changes for the BarlowTwins onto this PR directly.

**Comment by kashif** [comment]
Yes sure I will try to do that tomorrow if that’s ok?

**Comment by timoschowski** [comment]
> Yes sure I will try to do that tomorrow if that’s ok?

for sure, but there's no time pressure -- just wanted to make sure that you directly push to @vafl 's branch.

**Comment by kashif** [comment]
@vafl as far as I can see, all the losses can be now endowed with time series augmentations via the `RandomApply` on both representation branches. I still do not have an evaluation setup but if you do you can try it out.

**Comment by vafl** [comment]
> @vafl as far as I can see, all the losses can be now endowed with time series augmentations via the `RandomApply` on both representation branches. I still do not have an evaluation setup but if you do you can try it out.

yeah, the evaluation is tricky. We don't have a better setup than what is described in https://arxiv.org/abs/1901.10738 of using the embeddings for classification tasks with few labels.

File: src/gluonts/nursery/ts_embeddings/embed_model.py [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+        """
+        raise NotImplementedError()
+
+
+class ScalePreProcessor:
+    """
+    A pre-processor that takes a (possibly multi-variate) time series and normalizes each dimension by its mean
+    """
+
+    def __init__(self, out_dim, min_scale=1e-12):
```
Comment (mbohlkeschneider) [inline src/gluonts/nursery/ts_embeddings/embed_model.py:84]: Scaling by a very small scale can have weird side effects (blowing up of values). Maybe better to not scale at all in this case?

File: src/gluonts/nursery/ts_embeddings/embed_model.py [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+        agg_fns: List[str],
+        out_dim: int,
+        rescale: bool = True,
+        min_scale: float = 1.0e-12,
+    ):
+        self.agg_window = agg_window
+        self.rescale = rescale
+        self._agg = []
+        for agf in agg_fns:
+            m = re.match(r"q(\d\.\d+)", agf)
```
Comment (mbohlkeschneider) [inline src/gluonts/nursery/ts_embeddings/embed_model.py:126]: I absolutely do not get what this regex does. Maybe an example?

File: src/gluonts/nursery/ts_embeddings/pt_augmentation.py [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+        ones = torch.ones((x.shape[0] * x.shape[2]), device=x.device)
+        flip = torch.where(flip_index == 0, -ones, ones)
+
+        rotate_axis = np.arange(x.shape[2])
+        np.random.shuffle(rotate_axis)
+
+        return flip.reshape(x.shape[0], 1, x.shape[2]) * x[:, :, rotate_axis]
+
+
+class Permuation(nn.Module):
```
Comment (mbohlkeschneider) [inline src/gluonts/nursery/ts_embeddings/pt_augmentation.py:100]: typo

**Review by mbohlkeschneider — COMMENTED** [review]
Left some comments. To be honest, I tried my best here but this that has been growing too large to review it appropriately by a single person in a reasonable time. 

I suggest one of the two ways to move forward: 

1. Address only critical comments and leave todos so we can move forward with this and iteratively fix issues. 
2. Review different parts of this PR by different reviewers. 

Any other suggestions?

**Review by mbohlkeschneider — APPROVED** [review]
